text/trace/115-add-sample-to-span-end.md

```diff
@@ -0,0 +1,204 @@
+child spans that are created before the deferred span is. Examples here include
+a span to represent the receiving of data: the span that represents receiving data
+would have started before the name of the span is available, since that depends
+on receiving and parsing the incoming payload. The result will at worst be
+multiple root spans that should actually be children of the deferred span, since
+there will be no active span present when the children are created.
+
+### It is better to address the issues with sampling
+
+Given the need for an almost identical, most likely less robust code path
+for deferred spans, and the fact that it is a very common pattern, solving
+issues with sampling would ensure a more robust and maintainable code base.
+
+## Explanation
+
+## Additional Hooks
+
+A sampler would consume the same arguments it does today:
+
+- span name
+- span context
+- span kind
+- span attributes
+- span links
+
+The sampler would incorporate additional hooks, one to match each modification
+of events on an existing span:
+
+* onUpdateName, called by updateName
+* onSetAttribute, called by setAttribute
+* onAddEvent, called by addEvent
+* onSetStatus, called by setStatus
+* onEnd, called at the end of the span
+
+These methods will consume both the span, as well as the field or state that
+was changed. This enables optimizations in the case that samplers only depend
+on the data in the modification to make a decision, as well as more complex
+sample checks that may require all fields of the span.
+
+onEnd is also provided, as an optimization for samplers who may opt for
+running the sample decision after no further updates to the span are possible.
+
+## Samplers return shouldRetry
+
+Samplers will return an additional "shouldRetry" boolean value, that indicates
+whether the sample should be called again. Once shouldRetry has returned false,
+further sample hooks will not be called.
+
+shouldRetry is only valid when paired with a RECORD or RECORD_AND_SAMPLED result.
+A NOT_RECORD result results in no further calls to the sampler, regardless of
+the shouldRetry value.
+
+## Internal details
+
+## Trade-offs and mitigations
+
+### Rationale for separating shouldRetry from sampled result
+
+## Prior art and alternatives
+
+This implementation should address the following issues:
+
+- [remove warning on trace.updateName](https://github.com/open-telemetry/opentelemetry-specification/issues/468)
+- [remove warning on trace.updateName PR](https://github.com/open-telemetry/opentelemetry-specification/pull/506)
+- [Allow samplers to be called during different moments in the Span lifetime](https://github.com/open-telemetry/opentelemetry-specification/issues/307)
+- [Sampling decision is too late to gain much performance](https://github.com/open-telemetry/opentelemetry-specification/issues/620)
+
+### Tradeoffs of additional sample hooks
+
+- Pro: enables setting the recording decision early, which can skip additional
```

Comment:
I'm a bit confused. How does modifying the sampling decision during one of the hooks (let's say onEnd) set the recording decision early, compared to right now in where the decision is made upon creation?

Reply:
I would say this is a pro relative to other ways to address the problem above, rather than a pro vs the current solution.

To your point, the current spec sets the decision as early as possible.